library/Booster/Pattern/Util.hs

```diff
@@ -199,18 +202,29 @@ filterTermSymbols check = cata $ \case
         let concatSym = concatSymbol $ KListMeta def
             elemSym = klistElementSymbol def
             unitSym = unitSymbol $ KListMeta def
-         in if null heads
-                then [unitSym | check unitSym]
-                else filter check [concatSym, elemSym] <> concat heads
+         in case heads of
+                [] -> [unitSym | check unitSym]
+                [x] -> [elemSym | check elemSym] <> x
+                more -> filter check [concatSym, elemSym] <> concat more
     KListF def heads (Just (mid, tails)) ->
         let concatSym = concatSymbol $ KListMeta def
             elemSym = klistElementSymbol def
-            unitSym = unitSymbol $ KListMeta def
             ends = heads <> tails
          in mid
                 <> if null ends
                     then []
-                    else filter check [concatSym, elemSym, unitSym] <> concat ends
+                    else filter check [concatSym, elemSym] <> concat ends
+    KSetF def elements rest ->
+        let concatSym = concatSymbol $ KSetMeta def
+            unitSym = unitSymbol $ KSetMeta def
+            elemSym = klistElementSymbol def
+         in case elements of
+                [] -> fromMaybe [unitSym | check unitSym] rest
+                [single] ->
+                    single
+                        <> maybe [elemSym | check elemSym] (filter check [concatSym, elemSym] <>) rest
+                more ->
+                    filter check [concatSym, elemSym] <> fromMaybe [] rest <> concat more
```

Comment:
```suggestion
                [] -> fromMaybe [] rest
                [single] ->
                    single
                        <> maybe [] (concatSym:) rest
                more ->
                    [concatSym] <> fromMaybe [] rest <> concat more
```

Reply:
This is again because only `concatSym` is not total and the other two are?
I would like to keep the current code so it remains general. Otherwise we could specialise it to returning the partial symbols (i.e., hard-wire the `check` argument)  and rename `s/filterTermSymbols/partialTermSymbols`